### ISSUE TITLE
Cannot load auth plugin written for 0.9.3 in 0.10.0

### ISSUE THREAD

**Issue by bmcustodio** [emqtt/emqttd#254]
With the new plugin architecture, a custom-written authentication plugin that was working flawlessly under `0.9.3` stopped working in `0.10.0`. I modified the plugin so as to include the `etc/plugin.config` configuration file and imported it according to the wiki's instructions, but it just doesn't work. Do I need to modify anything in the Erlang code itself? Did the hooks change?

When I run `bin/emqttd console` I get

```
Hi!
18:39:34.583 [info] started Apps: [emqttd_auth_custom]
18:39:34.583 [info] load plugin emqttd_auth_custom successfully
```

being that `Hi!` is written by the plugin on `init/1`. However, when I try to connect to the broker, the authentication tokens that used to work do not work... It seems that the plugin isn't being used for authentication. How can I make it work?

**Comment by erylee** [comment]
Any error logs when authentication with your plugin?

**Comment by bmcustodio** [comment]
No, nothing is printed in the logs.

**Comment by bmcustodio** [comment]
Even if I use the following inside the plugin, nothing is printed when I
try to authenticate and the connection is refused:

    check(_, _, _) ->
      io: format("check was called"),
      ok.

**Comment by bmcustodio** [comment]
It seems like the plugin is loaded but not being called on auth.

**Comment by erylee** [comment]
Please check if the auth_mod is registered successfully. For example in emqttd_plugin_mysql:

```
ok = emqttd_access_control:register_mod(auth, emqttd_auth_mysql, {AuthSql, HashType}),
```

Or dump all the auth modules on console:

```
(emqttd@127.0.0.1)1> emqttd_access_control:lookup_mods(auth).
[{emqttd_auth_anonymous,[]}]
```

**Comment by bmcustodio** [comment]
OK, I will do so in a few hours time and post back the results. Thanks for
taking the time to investigate :-)

**Comment by bmcustodio** [comment]
Here's what I get from `emqttd_access_control:lookup_mods(auth).`:

```
(emqttd@127.0.0.1)2> emqttd_access_control:lookup_mods(auth).
[{emqttd_auth_custom,{state,<<"Test!">>}}]
```

The module seems to be loading OK. However, when I try to connect to the broker it doesn't seem to be called.

**Comment by erylee** [comment]
@bmcustodio, could you zip your plugin and send me by email if it does not contain confidential information?

**Comment by bmcustodio** [comment]
@erylee after a few hours struggling with this I managed to get it working. I'm still not quite sure what was wrong, though --- anyway, I changed the plugin's structure to match `emqttd_plugin_mysql` and it seems to be working.

The only thing that I'm missing is how to make the plugin active by default, as on a fresh build it will only become active after I run `emqttd_plugins:load(emqttd_auth_custom).`. Am I missing something?
